Network amount component

transducer opened the pull request:
fixes #13556

### Summary
Network Amount component as per [the design](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Mobile-Components?node-id=919%3A27533).

### Screenshot
![status-network-amount-ios](https://user-images.githubusercontent.com/5002921/186998400-6c4a9e27-13fc-48e0-a2e5-ab9762daa43d.png)

### Review notes
#### Icons
To create icons of various sizes I downloaded the SVG from Figma and used `rsvg-convert` version 2.54.4 and: 

```sh
rsvg-convert icon.svg -w 24 -h 24 > icon12@2x.png
rsvg-convert icon.svg -w 36 -h 36 > icon12@3x.png
```
Not sure though if the quality of the icons are correct using this approach.

To create icons of various sizes I made [this utility script](https://gist.github.com/transducer/b00ddf50c0d36fe30290e76df9757421), where`./convert.sh icon 12 16 20` creates three icon sizes using `rsvg-convert`.

#### Networks
I implemented switching between icons and names of networks (as in the the [Network Breakdown](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Mobile-Components?node-id=919%3A27526)), but not sure if that was needed.

Happy to process other review comments.

#### Platforms
- Android
- iOS

### Steps to test
- Open Status
- Enable new UI in advanced settings
- View sandbox

status: ready

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | c1d7156d | [#1](https://ci.status.im/job/status-mobile/job/prs/job/android-e2e/job/PR-13852/1/) | 2022-08-24 17:24:48 | ~8 min | `android-e2e` | [:package:`apk`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-220824-171647-c1d715-pr13852-x86.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-Mobile-220824-171647-c1d715-pr13852-x86.apk)|

ibrkhalil commented:
Re: [issue-13557](https://github.com/status-im/status-mobile/pull/13781)

Parveshdhull commented on src/quo2/components/network_amount.cljs:
```diff
@@ -0,0 +1,85 @@
+  {:icon {:small 12 :medium 16 :large 20}
+   :container {:small 32 :medium 32 :large 48}
+   :label {:small :label :medium :paragraph-2 :large :paragraph-1}
+   :amount-text {:small :paragraph-2 :medium :paragraph-1 :large :heading-2}
```
hi, why all these sizes? According to design it only has one size.

Parveshdhull commented on src/quo2/components/network_amount.cljs:
```diff
@@ -0,0 +1,85 @@
+            [status-im.i18n.i18n :as i18n]))
+
+(def ^:private sizes
+  {:icon {:small 12 :medium 16 :large 20}
+   :container {:small 32 :medium 32 :large 48}
+   :label {:small :label :medium :paragraph-2 :large :paragraph-1}
+   :amount-text {:small :paragraph-2 :medium :paragraph-1 :large :heading-2}
```
why all these sizes? According to design it only has one size.

transducer replied:
Thanks. The issue #13556 states "don't forget to add sizing to the issue". I assumed this meant to also create sizes. 

The `:small` size is now per the design, the other sizes I made equivalent to those of the [`quo2.components.group-avatar`](https://github.com/status-im/status-mobile/blob/develop/src/quo2/components/group_avatar.cljs#L7). I now see that the [Group Avatar design](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Mobile-Components?node-id=179%3A9760) indeed has different sizes, while the Network Amount has not. So "don't forget to add sizing to the issue" probably refers to something else and my assumption was mistaken.

Shall I remove `sizes` and hard code the size of the component as per the design?

Parveshdhull replied:
> don't forget to add sizing to the issue

Yes,I think it is different sizing. 
(@cammellos Please can you confirm if by sizing you meant complexity estimation?)

> Shall I remove sizes and hard code the size of the component as per the design?

Yes please

cammellos replied:
Yes, sorry for the confusion, I meant estimation

transducer replied:
Done, only small size now. Screenshot:

![status-network-amount-ios](https://user-images.githubusercontent.com/5002921/186896306-310e1c87-f790-42c6-9ade-1a59a077525f.png)

I am not sure about the quality of the icon. The [quo2.components.icon/icon](https://github.com/status-im/status-mobile/blob/develop/src/quo2/components/icon.cljs#L6) component makes it impossible to use a higher resolution image, since the provided size is appended to the icon name. [[issue-13557] Network breakdown component](https://github.com/status-im/status-mobile/pull/13781) uses bigger icons and shrinks them, using the old `icon` function. I thought about overloading the new `quo2.components.icon/icon` component so that a higher resolution image can be used, but not sure if that aligns with the reasoning behind the current implementation.

Also in [[issue-13557] Network breakdown component](https://github.com/status-im/status-mobile/pull/13781) the `@2x` and `@3x` icons are the same size. In this implementation they are following the convention that `@3x` is three times the size instead of two, but if the `@3x` icon is used the icon seems out of proportion for the network amount component and does not follow the design.

Can you let me know what the preferred approach is here?

Parveshdhull reviewed (commented):
Hi @transducer, Thank you very much for your contribution.

flexsurfer commented on src/quo2/screens/network_amount.cljs:
```diff
@@ -0,0 +1,70 @@
+            [reagent.core :as reagent]))
+
+(def ^:private networks
+  [#:network{:icon :main-icons2/ethereum
```
why `#:network` is needed ?

transducer replied:
I used namespaced keys to clarify it's a description of a network. Since it's only used as dummy data on this example screen it has no other benefits, so I will remove it.

transducer replied:
Changed.

flexsurfer commented on src/quo2/components/network_amount.cljs:
```diff
@@ -0,0 +1,85 @@
+    [divider size]]
+   [label size (i18n/label :t/on) network-name]])
+
+(defn view
```
it's better to name it `network-amount`

transducer replied:
I thought `view` is the somewhat better name, because `network-amount` reeked like a smurf naming convention to me. Naming it `network-amount/view` is not consistent with the other components though, and when requiring only the component and not the namespace, naming the component function also `network-amount` is clearer.

 I'll change it to `network-amount`.

transducer replied:
Done.

flexsurfer commented on src/quo2/components/network_amount.cljs:
```diff
@@ -0,0 +1,85 @@
+    :eth-value    1.2345678               ;; shown ETH value}"
+  [{:keys [size style network-name icon eth-value] :as _opts}]
+  [rounded-container size style
+   [row
```
why do we need a separate `row` component ?

flexsurfer replied:
and same question for `rounded-container`

flexsurfer replied:
it just makes code more complex and confusing, better if its readable in one form here

transducer replied:
It's following the single responsibility principle. In the long run I find it's more maintainable and readable to extract functions, so that every function does one thing only (but not too far of course, maybe I overdid it here). See, for example, the "social code" demonstrated in [Writing Quality Code in Erlang](https://youtu.be/CQyt9Vlkbis). I find it's more flexible, readable and maintainable that way. 

The downside is that to read the code for implementations you have to jump to the definitions (iff there is a need to know ). Advantages are that functions can be reused (e.g., `row` is used twice) and are encapsulated into clear names (e.g., the orchestrating component does not contain lower level implementation details like styling, but delegates this to lower level components like `rounded-container` and `row`, that could be extracted to a shared namespace).

If the preferred style is to create one component I am happy to change it. Also this component will be quite short.

I'll put everything (or maybe only the `rn/view`s) top level.

flexsurfer replied:
yes in theory it sounds good :)  but this component is so simple so its better to have all in one view

transducer replied:
Done.

flexsurfer commented:
hey @transducer could you please rebase and resolve conflicts , and also pls make sure `make test` running without errors locally, thanks

transducer commented:
@flexsurfer I rebased into one commit now, but there's still a conflict here. Locally there's none with latest develop. Any ideas how to fix?

transducer committed:
feat: network amount component and example screen (#13852)

- Switch between four example networks
- Add network icons
- Use i18n labels
- quo2.text for texts

transducer commented:
>hey @transducer could you please rebase and resolve conflicts , and also pls make sure make test running without errors locally, thanks 

@flexsurfer this is done now. I had some difficulty with intertwined history while rebasing 😅, maybe because I pulled develop a few times without rebase.

flexsurfer commented on src/quo2/screens/network_amount.cljs:
```diff
@@ -0,0 +1,52 @@
+
+(defn- state->opts [{:keys [eth-value network]
+                     :or {eth-value 5.123456}
+                     :as _state}]
```
_state is unused

transducer replied:
@flexsurfer this is pure for readability. The `_`-prefix indicates the variable is unused, as suggested by clj-kondo and described in the [Clojure styleguide - unused bindings](https://github.com/bbatsov/clojure-style-guide#unused-bindings). Because the name is `state->opts` it's clear the input is `state`, but I think it's still a good practice to name destructured maps. If you prefer to have this removed let me know.

flexsurfer replied:
but an alias is optional, and if it's not used, why do we still want to keep it?

flexsurfer replied:
it's not a big deal :) just wondering, seems like just individual preferences

transducer replied:
Not a big deal but I removed it anyway because this was only the second contribution where `:as _` was used in this repository, and it does not contribute that much in this case since the function name is clear.

flexsurfer commented:
duplicate https://github.com/status-im/status-mobile/pull/13781

transducer commented:
@flexsurfer although similar, this is not a duplicate

flexsurfer commented:
сс @ibrkhalil

flexsurfer commented:
@ibrkhalil should we merge this PR then and you can reuse it in your PR ?

flexsurfer commented:
thank you for your effort @transducer , we'll cherry-pick your commit into another PR
